### ISSUE TITLE
git blame is slow

### ISSUE THREAD

**Issue by billhoffman** [libgit2/libgit2#3027]
We are using gitlab and found that with a large repository with a long history, the git blame was timing out.  After some digging, we found that the slowness comes from libgit2.

If you build the libgit2 example blame and run it on some of our files it takes orders of magnitude longer than command line git.

For example:

  time git blame CMakeLists.txt
  real    0m0.761s
  user    0m0.607s
  sys     0m0.124s

  blame CMakeLists.txt > foo

  real    0m37.881s
  user    0m0.000s
  sys     0m0.046s

Is this a known issue?

I tested with this version:
git describe
v0.16.0-6044-g92a8194

**Comment by ethomson** [comment]
April fools?

No, but seriously, that's a lot slower.  I don't know why that might be the case offhand (I don't use libgit2's blame, I don't think that anybody's actually used it in production.)

Regrettably, the contributor who brought blame into libgit2 is not working on the project anymore, so this doesn't have much of a sponsor (per se).  Blame did come in fairly hot - and I suspect that the initial revision of it was aiming for correctness, not necessarily performance.  I suspect that there may be some low hanging fruit that could be plucked.

**Comment by arrbee** [comment]
I think you may be comparing a Debug build of libgit2 against Git. Running with a libgit2 release build, I see:

```bash
$ time git blame CMakeLists.txt > /dev/null

real	0m0.128s
user	0m0.096s
sys	0m0.026s

$ time ./build/examples/blame CMakeLists.txt > /dev/null

real	0m1.704s
user	0m1.602s
sys	0m0.100s
```

That is still an order of magnitude off, but seems to be about 4 or 5 times faster than you're seeing.

That being said, libgit2's blame code is largely derived from core Git and one of the authors of some of the recent optimizations to Git's blame code has explicitly asked this his work not be included in libgit2.

If the GitLab folks wanted to look into profiling and optimizing libgit2's blame code, that would be wonderful!

**Comment by arrbee** [comment]
(In case there is confusion, I realize we're using different machines - when I write "4 or 5 times faster" I just meant that in your example the libgit2 speed was about 50x slower than Git and in mine it was "only" about 12x slower.)

**Comment by peff** [comment]
You can also try comparing git v2.0 to v2.1. That includes the optimizations from github/git@7e6ac6e4391caa0fc379cb699013d503380e4214 that @arrbee mentioned. On my machine it does not make much of a difference on `CMakeLists` though (it makes more of a difference when there are a lot of separate chunks to blame). So that may not be relevant here.

**Comment by billhoffman** [comment]
Not April fools...  :)

I did just build release.  Note, I was running on Windows, that could be the slow down.  However, on Linux this file is particularly slow:
https://gitlab.com/jake-stookey/kw-gh-test/blob/master/CMake/CTestCustom.cmake.in

If you clone VTK on a linux machine, and compare times it is pretty bad:

Also, found on github:
https://github.com/Kitware/VTK/blame/master/CMake/CTestCustom.cmake.in

time git blame CMake/CTestCustom.cmake.in > out
real	0m0.434s
user	0m0.346s
sys	0m0.086s

 

time ~/libgit2/build/examples/blame CMake/CTestCustom.cmake.in > out
real	0m28.209s
user	0m24.359s
sys	0m3.156s

So, that was a release build of libgit2.

**Comment by Therzok** [comment]
Kicking this dead horse to life.

Found out that [blame like git](https://github.com/libgit2/libgit2/blob/c69c042e0e1eba787cd25d7c5c98d45b0ae895c3/src/blame_git.c#L440) is always doing rename detection, regardless of what's happening.

Probably this is a _huge_ factor of slowdown?

**Comment by xuyungit** [comment]
Yes, 99% time would  be consumed by the function find_origin, but the hot spots are
git_diff_tree_to_tree
git_commit_tree
gi_blame__get_origin
instead of git_diff_find_similar
So I think it's not related to rename detection.

**Comment by carlosmn** [comment]
That shows that it is related, `git_diff_tree_to_tree()` is what you would have to call a lot in order to perform rename detection. There's also the measurements which show that this does speed up when you turn off those options.

**Comment by Therzok** [comment]
@xuyungit See https://github.com/libgit2/libgit2/pull/3167 for reference

**Comment by Mange** [comment]
> That being said, libgit2's blame code is largely derived from core Git and one of the authors of some of the recent optimizations to Git's blame code has explicitly asked this his work not be included in libgit2.

I am very curious to find out why they said this. Is there someway for me to read their motivations somewhere? I cannot find anything on Google, but I might be using the wrong keywords.
Most OSS happen in the open, especially the Linux teams with their mailing lists but this could've been a private correspondence too.

**Comment by peff** [comment]
@Mange The original thread is at https://public-inbox.org/git/1398470210-28746-1-git-send-email-dak@gnu.org/ if you're interested.

**Comment by Mange** [comment]
I came in to comment about not finding it in the original link, but that I finally found it by searching those archives. I missed your edit. :-)

**tl;dr** for others: The original contributor does not wish to contribute to codebases without an unmodified version of the GPL for philosophical reasons, unless greatly compensated.

Seems like this part of the `COPYING` file is the problem:

```
----------------------------------------------------------------------

			LINKING EXCEPTION

 In addition to the permissions in the GNU General Public License,
 the authors give you unlimited permission to link the compiled
 version of this library into combinations with other programs,
 and to distribute those combinations without any restriction
 coming from the use of this file.  (The General Public License
 restrictions do apply in other respects; for example, they cover
 modification of the file, and distribution when not linked into
 a combined executable.)

----------------------------------------------------------------------
```

However, the original message chain about the patches mention that JGit (a Java implementation) does these things already, so perhaps code could be based on that codebase instead? Just an idea.
See this thread for the details: https://public-inbox.org/git/CAJo=hJukmej1rJXuVoECwd7AxmSue8Wmv4rBmCHEYcWBWNarSw@mail.gmail.com/

**Comment by ethomson** [comment]
I don't think that will help - blame/annotate is a well solved problem, and jgit's internals are sufficiently different that it's not a straight port.  Even taking git's blame code is unlikely to be particularly straightforward.  That's, in fact, how we got the code we have now.

Really somebody just needs to throw out the code we have and start over.  But nobody is sufficiently inclined to do that.

**Comment by wtdcode** [comment]
Hello, any update on this thread?

**Comment by pks-t** [comment]
> Hello, any update on this thread?

Unfortunately not. Nobody has stepped forward yet with fixes to speed this up (I think?) and we're unfortunately lacking the time to do this.

**Comment by qaqland** [comment]
update:

Facing large repositories with many files, the speed is still very slow. There is no good way to improve `git_diff_tree_to_tree()`.

For reference, GitLab(gitaly) and Gitea use git command get CommitDiff:

- https://gitlab.com/gitlab-org/gitaly/-/blob/3f33592a35304837e020da8d3cdc61f585ab4692/internal/gitaly/service/diff/commit_delta.go#L37
- https://github.com/go-gitea/gitea/blob/abe743df797c862ec563610470121c71176d892a/services/gitdiff/gitdiff.go#L1132

**Comment by AHSauge** [comment]
I dug a bit into this, and as has been mentioned earlier here, rename detection is a major contributor to the performance due to an NxM complexity comparison in `git_diff_find_similar` (technically it's limited to Nx1000). Sacrificing some accuracy, I'm able to more than half the time blame takes with these two changes in `similarity_measure` (found in `diff_tform.c`)

1. Reduce the cut-off from 8x to 2x file size. Assumption here being that even doubling the file size is already enough to not regard it as a rename
2. Reject comparison between files where the basename (excluding file extension) is different. The assumption here being that renames are predominantly going to be moving files or changing the file extension.

On my computer, these changes makes things go from
```
time ./build-release/examples/lg2 blame CMakeLists.txt > /dev/null

real    0m2,576s
user    0m2,503s
sys     0m0,060s
```

to 
```
time ./build-release/examples/lg2 blame CMakeLists.txt > /dev/null

real    0m1,171s
user    0m1,118s
sys     0m0,046s
```

Big improvement, but still of course nowhere near the git CLI
```
time git blame CMakeLists.txt > /dev/null

real    0m0,052s
user    0m0,043s
sys     0m0,009s
```

I'm not necessarily saying these are valid assumptions or changes, but this does suggest that improved filtering and rejection is required to improve performance. That said, it's not the entire story here. I've not had time to debug this further, but forcing `exact_match` to be true in `similarity_measure` results in
```
ime ./build-release/examples/lg2 blame CMakeLists.txt > /dev/null

real    0m0,887s
user    0m0,818s
sys     0m0,064s
```
meaning performance must be suffering due to other sources as well.
